### PR TITLE
Run only activated rules from the profile

### PR DESCRIPTION
**Pull request by saberduck**
Fixes  https://github.com/SonarSource/SonarTS/issues/111 
Contributes to https://github.com/SonarSource/SonarTS/issues/103

### CODE REVIEW

File: sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java [hunk 1]
```diff
@@ -65,10 +71,11 @@
   /**
    * ExecutableBundleFactory is injected for testability purposes
    */
-  public ExternalTypescriptSensor(ExecutableBundleFactory executableBundleFactory, NoSonarFilter noSonarFilter, FileLinesContextFactory fileLinesContextFactory) {
+  public ExternalTypescriptSensor(ExecutableBundleFactory executableBundleFactory, NoSonarFilter noSonarFilter, FileLinesContextFactory fileLinesContextFactory, CheckFactory checkFactory) {
```
Comment (SonarTech) [inline sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java]: MINOR Split this 189 characters long line (which is greater than 180 authorized). [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00103)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [rules](https://github.com/SonarSource/SonarTS-Plugin/blob/5f9961d7ba8205eaed41e993b0f47a8dd2c90ab5/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/rules): Add a 'package-info.java' file to document the 'rules' package [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1228)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/SonarTSCoreBundleTest.java [hunk 2]
```diff
@@ -71,4 +88,26 @@ public void should_fail_when_bad_zip() throws Exception {
...
   }
 
+  @Test
+  public void should_activate_rules() throws Exception {
+    ExecutableBundle bundle = new SonarTSCoreBundleFactory("/testBundle.zip").createAndDeploy(DEPLOY_DESTINATION);
+    Rules rules = new Rules(new TestRule());
+    bundle.activateRules(rules);
+    List<String> strings = Files.readLines(new File(DEPLOY_DESTINATION, "sonarts-core/tslint.json"), StandardCharsets.UTF_8);
+    String json = strings.stream().collect(Collectors.joining());
+    assertThat(json).contains("\"test-rule\": true");
```
Comment (vilchik-elena) [inline sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/SonarTSCoreBundleTest.java]: @saberduck can we assert that it's the only rule in config?

File: sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/rules/RulesTest.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    return activeRules;
+  }
+
+  private ActiveRule mockActiveRule(String key) {
+    ActiveRule activeRule = mock(ActiveRule.class);
+    when(activeRule.ruleKey()).thenReturn(RuleKey.of(TypeScriptRulesDefinition.REPOSITORY_KEY, key));
+    return activeRule;
+  }
+
+  private static class ConfigurableTypeScriptRule extends org.sonar.plugin.typescript.rules.ConfigurableTypeScriptRule {
```
Comment (vilchik-elena) [inline sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/rules/RulesTest.java]: @saberduck why not think of another name?

File: sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java [hunk 4]
```diff
@@ -227,6 +241,7 @@ private void saveHighlights(SensorContext sensorContext, Highlight[] highlights,
   }
 
   private static class Failure {
+    String failure;
```
Comment (SonarTech) [inline sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java]: MAJOR Rename field "failure" [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1700)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by vilchik-elena** [comment]
Manually tested

File: sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/rules/RulesTest.java (hunk 3 above)
Comment (saberduck) [inline sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/rules/RulesTest.java]: In fact, after moving tslint key into the rule class itself, there is no reason to keep this class at all. The utility method can be moved to the `TypeScriptRule` interface

File: sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java (hunk 4 above)
Comment (vilchik-elena) [inline sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java]: won't fix

File: sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/rules/Rules.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    .put("no-unconditional-jump", "S1751")
+    .put("no-useless-increment", "S2123")
+    .put("no-use-of-empty-return-value", "S3699")
+    .put("no-variable-usage-before-declaration", "S1526")
+    .build();
+
+  private final List<TypeScriptRule> ruleInstances;
+
+  @VisibleForTesting
+  public Rules(TypeScriptRule... rules) {
```
Comment (saberduck) [inline sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/rules/Rules.java]: I removed the whole constructor and reworked the tests

File: sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/rules/TypeScriptRules.java [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.BiMap;
+import com.google.common.collect.ImmutableBiMap;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableList.Builder;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonPrimitive;
+import java.util.ArrayList;
+import java.util.Arrays;
```
Comment (SonarTech) [inline sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/rules/TypeScriptRules.java]: MINOR Remove this unused import 'java.util.Arrays'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by vilchik-elena** [comment]
@saberduck Feel free to merge

File: sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/SonarTSCoreBundleTest.java (hunk 2 above)
Comment (saberduck) [inline sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/SonarTSCoreBundleTest.java]: I will assert content of the whole file instead, that way we are sure.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Pull request merged**